### ISSUE TITLE
Inconsistent training results

### ISSUE THREAD

**Issue by gitlonglong** [lmy98129/VLPD#7]
Hi, is the public code final? Why I installed the same environment and used the same random seed to train, but the usage result can't reach so good?

Epoch 300 end, AvgLoss is 0.016301, Time used 153.0sec.
Epoch 287 has lowest loss: 0.0151585
Epoch 208 has lowest MR: 0.1006502
Val: epoch_num: 300 loss: 0.0163009 Summerize: [Reasonable: 10.52%], [Reasonable_small: 13.92%], [Reasonable_occ=heavy: 36.63%], [All: 36.09%], lr: 0.000200
Epoch 208 has lowest MR: 0.1006502

I didn't change the settings in the config.py file, except to change the path. I also noticed that the weights file you provided is called "VLPD-262.pth", but the weights file I got from the training is saved as "CSP-208.pth". Each run should come out the same, if the random seed is not replaced. Are there other modifications needed or details I'm missing? 
Looking forward to your reply?

**Comment by lmy98129** [comment]
Hello, this public code is final.

**Comment by lmy98129** [comment]
Hello, 

This public code is surely final, if it is ensured to be installed according to the instructions in README. 

The main issue might be "The same random seed does not always works across all the machines, try change another one.", which is due to the PyTorch: https://github.com/pytorch/pytorch/issues/38219

For better reproduction, please follow our further instruction to try better random seeds:
https://github.com/lmy98129/VLPD/blob/main/docs/Others.md#about-reproducibility
Meanwhile, the tensorflow version of our baseline CSP is also greatly affected by randomness. Please refer to their trial tables for each dataset:
https://github.com/liuwei16/CSP?tab=readme-ov-file#models

The saving file name is mannually changed from "CSP" to "VLPD" for distingushing from the baseline during the experiments. Because we did not change the class name ``net.module.__class__.__name__`` from "CSP" to "VLPD". Now that there are differences between machines, the best epoch number will differ as well:
https://github.com/lmy98129/VLPD/blob/main/trainval_distributed.py#L161

Mengyin Liu

**Comment by gitlonglong** [comment]
> Hello,
> 
> This public code is surely final, if it is ensured to be installed according to the instructions in README.
> 
> The main issue might be "The same random seed does not always works across all the machines, try change another one.", which is due to the PyTorch: [pytorch/pytorch#38219](https://github.com/pytorch/pytorch/issues/38219)
> 
> For better reproduction, please follow our further instruction to try better random seeds: https://github.com/lmy98129/VLPD/blob/main/docs/Others.md#about-reproducibility Meanwhile, the tensorflow version of our baseline CSP is also greatly affected by randomness. Please refer to their trial tables for each dataset: https://github.com/liuwei16/CSP?tab=readme-ov-file#models
> 
> The saving file name is mannually changed from "CSP" to "VLPD" for distingushing from the baseline during the experiments. Because we did not change the class name `net.module.__class__.__name__` from "CSP" to "VLPD". Now that there are differences between machines, the best epoch number will differ as well: https://github.com/lmy98129/VLPD/blob/main/trainval_distributed.py#L161
> 
> Mengyin Liu

Thanks for the reply, I found that changing the environment didn't affect the results, but changing the GPU did. I tried it with A40 and A100 respectively (no available 3090 at the moment) and also changed the random seed a couple of times and the best results I've run so far are as follows:

Epoch 300 end, AvgLoss is 0.016232, Time used 151.0sec.
Epoch 289 has lowest loss: 0.0152953
Epoch 296 has lowest MR: 0.0974612
Val: epoch_num: 300 loss: 0.0162318 Summerize: [Reasonable: 10.37%], [Reasonable_small: 13.12%], [Reasonable_occ=heavy: 36.95%], [All: 35.42%], lr: 0.000200
Epoch 296 has lowest MR: 0.0974612

It does seem that the results of the experiment are highly randomized. Thanks again for your response and for your meaningful paper.

**Comment by lmy98129** [comment]
@gitlonglong 

Thanks for your responses, especially the detailed analysis on different environments and GPU platforms, which will assist us to tackle the reproduction problems due to the randomness in the future.

If you have further questions or suggestions about this work, please do not hesitate to contact us!

Mengyin Liu
